ensure we delete individual submissions on disk

redshiftzero opened the pull request:
# Description

Fixes #892.

# Test Plan

Given the importance of the deletion functionality, I recommend testing the following cases:

**_Precondition for each case:_** Via the source interface, create a source with at least one file. Log into the client and download the file.

1. Source is deleted by the local user. The entire source directory `~/.securedrop_client/data/intrusive_smattering` should be deleted. 
2. Source is deleted by another user (you can simulate this by deleting an entire source in the JI). The entire source directory `~/.securedrop_client/data/intrusive_smattering` should be deleted. 
3. An individual submission is deleted by another user (you can do this by deleting an individual submission in the JI). The file _and_ its enclosing folder should be deleted: `~/.securedrop_client/data/intrusive_smattering/1-intrusive_smattering-doc`. 

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [x] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [ ] These changes should not need testing in Qubes

If these changes add or remove files other than client code, packaging logic (e.g., the AppArmor profile) may need to be updated. Please check as applicable:

 - [ ] I have submitted a separate PR to the [packaging repo](https://github.com/freedomofpress/securedrop-debian-packaging)
 - [x] No update to the packaging logic (e.g., AppArmor profile) is required for these changes - I ran through the test plan above using the existing AppArmor profile
 - [ ] I don't know and would appreciate guidance

rmol commented on securedrop_client/storage.py:
```diff
@@ -544,21 +540,26 @@ def delete_source_collection(journalist_filename: str, data_dir: str) -> None:
+        # When a file is being processed (e.g. its filename is being extracted), it
+        # is stored in an intermediate state in a folder in the source data dir.
+        # We should also delete this entire directory.
+        in_progress_folder = os.path.join(
```
The name `in_progress_folder` seems inaccurate, as it exists when the `File` has been fully downloaded and decrypted.

rmol replied:
And it might be more future-proof to use `os.path.dirname(obj_db.location(data_dir))` here, in case we change the paths in `db.py`.

rmol commented on tests/test_storage.py:
```diff
@@ -1092,6 +1099,53 @@ def test_delete_single_submission_or_reply_race_guard(homedir, mocker):
+    test_obj = db.File(
+        source=source, uuid="test", size=123, filename=file_server_filename,
+        download_url='http://test/test')
+    source_directory = os.path.join(
```
Here again you could use `File.location`, unless you want this to be a canary for changes to that function. :slightly_smiling_face:

redshiftzero commented:
thanks for the comments! addressed and repushed. With respect to the two other findings, I believe they are covered by two other bugs: the conversation refresh not updating as part of #891, and the snippets not updating in #922

rmol commented:
Changes look good, can be merged after a rebase.

redshiftzero commented:
thanks for review! rebased

rmol commented:
Oh, bad news. I think some of the changes on master have broken things.

1. [x] Source is deleted by the local user. The entire source directory `~/.securedrop_client/data/intrusive_smattering` should be deleted. 

This worked, but the new "pending deletion" state wrote "Deleting..." over the empty conversation view text. It eventually corrected of course, but it was ugly in the meantime.

2. Source is deleted by another user (you can simulate this by deleting an entire source in the JI). The entire source directory `~/.securedrop_client/data/intrusive_smattering` should be deleted. 

This crashed the client with:
```
Traceback (most recent call last):
  File "/home/user/src/fpf/securedrop-client/securedrop_client/logic.py", line 459, in on_sync_success
    self.update_sources()
  File "/home/user/src/fpf/securedrop-client/securedrop_client/logic.py", line 498, in update_sources
    self.gui.show_sources(sources)
  File "/home/user/src/fpf/securedrop-client/securedrop_client/gui/main.py", line 150, in show_sources
    self.main_view.show_sources(sources)
  File "/home/user/src/fpf/securedrop-client/securedrop_client/gui/widgets.py", line 695, in show_sources
    deleted_sources = self.source_list.update(sources)
  File "/home/user/src/fpf/securedrop-client/securedrop_client/gui/widgets.py", line 926, in update
    del self.source_widgets[list_widget.source_uuid]
KeyError: 'da31ba2b-020e-4339-82d3-7ffa7858e47a'
```
3. An individual submission is deleted by another user (you can do this by deleting an individual submission in the JI). The file _and_ its enclosing folder should be deleted: `~/.securedrop_client/data/intrusive_smattering/1-intrusive_smattering-doc`. 

This went fine, except for the previously noted bugs (#891, #922).

redshiftzero commented:
Ahh ok, thanks for flagging. OK I'll need to take another look at this on Monday to see what happened and fix.

redshiftzero commented:
> the new "pending deletion" state wrote "Deleting..." over the empty conversation view text

I believe this one is #929 (recent bug on master)

Otherwise for the traceback you found, I was _not_ able to reproduce, however I added a guard to ensure that we gracefully handle if the uuid of the source widget is missing from the `source_widgets` dict (what's happening in your traceback afaict). 

Rebased/retested/pushed on latest master (8ab61016d9d10bfbeccd3e1824a3bac404b4900c)

redshiftzero commented:
Rebased/retested on latest `master` (a4a69af7ad44a841fcfb08f4cb0d69ad5b4a9f33)
